query-algebrizer/src/types.rs

```diff
@@ -194,6 +194,88 @@ pub enum ColumnConstraint {
     HasType(TableAlias, ValueType),
 }
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum ColumnConstraintOrAlternation {
+    Constraint(ColumnConstraint),
+    Alternation(ColumnAlternation),
+}
+
+impl From<ColumnConstraint> for ColumnConstraintOrAlternation {
+    fn from(thing: ColumnConstraint) -> Self {
+        ColumnConstraintOrAlternation::Constraint(thing)
+    }
+}
+
+/// An ordered collection of constraints that must all match.
+#[derive(PartialEq, Eq)]
+pub struct ColumnIntersection(pub Vec<ColumnConstraintOrAlternation>);
```

Comment:
Yes, this is deliberate — another way of phrasing your point is "these types aren't recursive", which at present is exactly what I want. This is enough to very simply implement simple `or`, and it won't be painful to extend this to `not`. I doubt we'll need more than that.